### DISCUSSION TITLE
Dead code linter

### DISCUSSION THREAD

**Discussion by dikovitskys** [MobileNativeFoundation/discussions#156] [General]
Hi everyone!

I would like to start a discussion regarding dead code linters and your company's best practices. Currently at Uber we use [SwiftCodeSan](https://github.com/uber/swiftcodesan) - our own implementation that was open-sourced but haven't been updated in a while since engineer that worked on it left the company. Right now we see two main routes:

- continue working on it
- find alternative open-source solution

Please share your experience and approaches for this matter, thanks!

P.S. Since previously I worked at Lyft and saw its dead code linter at work (and Lyft uses bazel as well), I would like to hear more from Lyft engineers in particular - @keith @sberrevoets 🙂

**Comment by keith** [comment]
We currently use an internal tool that uses the index store to detect dead code. There are a lot of details but in general I think it's a super solid approach that @jpsim wrote

**Reply by jpsim** [reply]
I'll share a bit of my experience at Lyft with tools to detect dead Swift code.

## Initial approach (SwiftLint)

For about 2 years, we ran SwiftLint's [unused_declaration](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/Lint/UnusedDeclarationRule.swift) analyzer rule once daily on our entire iOS codebase, the CI job posted the results to a Slack channel where volunteers inspected the results and manually made PRs to remove dead code. This was tedious, caused delays where we'd have to chase down code owners, and involved context switches for them. Plus we found that code was often unused _by mistake_, meaning that bugs were shipping due to this delay.

SwiftLint's analyzer rules require a complete and accurate [compilation database](https://clang.llvm.org/docs/JSONCompilationDatabase.html) and executes SourceKit requests to compute all the declarations and references. This provides very accurate results, but is extremely slow. For Lyft's iOS monorepo of 1.3M lines of Swift code, a full lint took 3 hours on CI.

This is a decent solution if your project is small (<100k LOC) or if you can shard+parallelize the job.

## New approach

Last year I wrote a new internal tool applying the lessons from SwiftLint's rule but avoiding the SourceKit requests, which are responsible for making SwiftLint's analyzer rules so slow.

This tool is written in ~1k lines of Swift code, and processes Lyft's 1.3M lines of Swift code in less than 30 seconds.

It requires a complete [index store database](https://docs.google.com/document/d/1cH2sTpgSnJZCkZtJl1aY-rzy4uGPcrI-6RrUpdATO2Q/edit), and producing that makes up the bulk of the CI job time, usually taking 3-8 minutes depending on the change.

At a high level, the deadcode tool does the following:

1. Collect units & records from the index store database
2. Generate the protocol conformance graph
3. Collect declarations & references
4. Compute unused declarations

The tool reads data from the index store, also runs some [SwiftSyntax](https://github.com/apple/swift-syntax) visitors, generates a protocol & inheritance graph (to detect code that's accessible via indirection), then collects declarations & references (applying some exceptions) and finally computes unused declarations (also applying some exceptions).

## Open sourcing

Although there are no concrete plans to open source, the tool was built with that eventual goal.

In my ideal world, I'd really want to replace SwiftLint's `unused_declaration` rule with this new implementation, but I need to sort through the following:

1. Either remove the dependency on an internal library for accessing index store contents or open source it
2. Go through Lyft's process of open sourcing code
3. Fix the SwiftSyntax distribution issues that we've run into with SwiftLint in the past

None of these are necessarily hard but they do take time and this work falls outside my job's typical responsibilities so I haven't prioritized doing any of that.

**Reply by rcrahul43** [reply]
@jpsim Any update to open source this tool. We would like to give it a try and also contribute if required.

**Comment by mgray88** [comment]
I know of [Periphery](https://github.com/peripheryapp/periphery). I haven't been able to try it out yet though

**Reply by scottluxenberg** [reply]
We've done some usage/analysis with Periphery, though it does have some issues dealing with Obj-C code we'd observed

**Reply by serjooo** [reply]
We have used Periphery at HelloFresh and it has been quiet useful and reliable. Would definitely recommend it!

**Reply by keith** [reply]
I haven't tried periphery in a while but last I did it had a significant number of false positives in our project. I'm not sure at this point if it's too tied to xcodebuild for me to be able to try again

**Reply by ileitch** [reply]
You can point directly at an index store and instruct it to skip the build step. But it's still tied to xcodebuild or SPM for identifying the targets to scan.

I'd be very interested in hearing more about your use case. My hope is that Periphery can be used by projects of all sizes.

**Reply by keith** [reply]
Ah great to know. Yea I think for these "outside of the norm" cases we'd probably need to provide some mapping of file <-> target for you, maybe in some json format or something. I think everyone from the bazel / buck world like ourselves would be in that case

**Reply by ileitch** [reply]
@keith Not sure if you're still interested, but Periphery now supports providing a file -> targets mapping file: https://github.com/peripheryapp/periphery#build-systems. Let me know if you give it a try, I'm happy to address any issues/questions.

**Reply by keith** [reply]
thanks! hopefully I can find some time to try it out again

**Reply by tinder-maxwellelliott** [reply]
.

**Reply by maxwellE** [reply]
For those interested I [put together a POC of this working with Bazel](https://github.com/bazelbuild/rules_swift/commit/e5e5952dd0023d7304b1899337231bc1db827df0). Should be simple enough to get anyone up and running with Periphery

**Reply by jpsim** [reply]
I took Periphery for a spin last week in Lyft's iOS codebase, which is already already pretty heavily culled for dead code using our internal tool, and it caught 3 categories of unused declarations that we previously were ignoring:

* Enum cases: skipped because enums with raw representable associated values can be created indirectly
* Initializers: skipped because the Swift compiler fails to register some references to initializers in some cases, such as when invoking an initializer with `Self(a:b:)` instead of `MyType(a:b:)`
* Parameters: skipped because this is extremely common mostly for protocol conformance and subclassing

Awesome! Together these checks flagged 3.3k LOC across 700 files of dead code in our codebase of 1.6M lines of Swift code.

This motivated me to extend our internal tool to catch the first two categories (enum cases & initializers), the third one I haven't cracked yet.

You might ask yourself, so if Periphery catches legitimate dead code that our internal tool misses, why not switch to using it instead of our in-house tool? There are a few reasons:

1. Speed: Periphery took ~9 minutes on an M1 Max to analyze our codebase, whereas our tool takes less than 30 seconds.
2. False positives: There were a few hundred false positives whereas we have very few with our tool.
3. Since we generally don't allow dead code to be checked in to our codebase, there's no easy way to identify code that our internal tool flags as unused that Periphery would have missed.

I've been "preparing the way" to open source our tool as a SwiftLint rule (WIP attempt in https://github.com/realm/SwiftLint/commit/05bc596daeec87b1fa72cdf85a568bc9091d4b91) but don't have an ETA on when that'll be in a shippable state just yet.

Kudos @ileitch on building a really awesome tool, that's by far and away the best option publicly available today. Once Lyft's tool is publicly available we should compare notes and see what we can learn from each other to improve our respective tools. I know you'd probably like more details on the false positives I found as well as more details on the performance profiles, I hope to be able to take the time to collect and share that information with you in the near future.

**Reply by ileitch** [reply]
Thanks for giving it a try, @jpsim!

> 1. Speed: Periphery took ~9 minutes on an M1 Max to analyze our codebase, whereas our tool takes less than 30 seconds.

I'm sure there's some low-hanging fruit to address here. I've not spent too much time profiling, to be honest. I hope to integrate Periphery in Reddit's build system this year, so I'll take a closer look at performance then. One possible contributing factor is that Periphery has a few more analysis techniques than your internal tool, those don't come for free.

> 2. False positives: There were a few hundred false positives whereas we have very few with our tool.

I try to fix all false positives as soon as they're reported, but unfortunately, some of them are caused by [Swift bugs](https://github.com/peripheryapp/periphery#known-bugs). I've implemented workarounds for some of them, but others aren't so easy. I'm hoping I can put some time aside to address these @Reddit. I'd *love* to get more details on the false positives you're seeing.

> Kudos @ileitch on building a really awesome tool, that's by far and away the best option publicly available today.

Many thanks!

**Reply by ileitch** [reply]
Hey @jpsim  just wanted to let you know that the latest release ([2.12.2](https://github.com/peripheryapp/periphery/releases/tag/2.12.2)) fixes a handful of false positives and performance issues. On an M1 Max 64GB I'm able to scan 369 targets, ~7200 files, and ~750K LOC in 22 seconds. There are still more optimizations I plan to make, but for now, I think I've fixed the most egregious bottlenecks.

If you're willing to give it another shot, I'd love to hear if this release performs any better for you.

A couple of notes on performance issues:
1. The index/report-exclude mechanism was performing a _lot_ of unnecessary disk reads, so on a large project this could have easily chewed up multiple minutes. This is now fixed, and it no longer touches the disk at all. Instead, it uses [https://github.com/ileitch/swift-filename-matcher](https://github.com/ileitch/swift-filename-matcher) for glob matching in memory.
2. If you integrated Periphery with SwiftPM, just be aware that you may have been inadvertently using a debug build.

**Reply by ThePragmaticArt** [reply]
Periphery definitely has its quirks, but was the best tool we could find give or take 1-2 years ago at Ford Motors.

It definitely had a few performance issues (as noted above), but generally caught what we needed it to albeit a few false positives.

I have yet to find a better tool, but would love any alternatives that are; faster, more reliable with Objective-C code, have a more intuitive way of handling libraries, and less false positives.

**Reply by ileitch** [reply]
@ThePragmaticArt I'd love to take a look at those false positives if you can file them https://github.com/peripheryapp/periphery/issues.

> more reliable with Objective-C code

I'm not aware of any objc specific false positives at the moment. Also worth noting is that I added the `--retain-objc-annotated` option recently to only retain declarations explicitly marked with `@objc` instead of everything potentially accessible to the objc runtime. The benefit is the detection of more unused code, with the caveat that you'll need to apply `@objc` on public NSObject types still used in objc (which is probably good practice anyway).

> a more intuitive way of handling libraries

Could you expand more on this, please?

**Reply by ThePragmaticArt** [reply]
> > a more intuitive way of handling libraries
> 
> Could you expand more on this, please?

Given the project structure, we were required to maintain a periphery-targets.yml file with a list of 100+ targets within the dependency graph in the project in order to run against the overarching mono-repo. This was quite cumbersome to say the least. We did eventually get to a point where generating this file was automated through a project generator or series of scripts, but never the less it would have been nice to have an accurate way of building that graph based on a primary target while periphery picked up cascading target dependencies.

This was a little while back and mostly challenges I'm pulling from past memory. It's possible things have changed, I may have to dig into a sample and re-try.

**Reply by maxwellE** [reply]
I have found periphery very simple to integrate using a Bazel aspect, I wonder if direct Bazel support could be added to periphery via a ruleset?

**Reply by ileitch** [reply]
@ThePragmaticArt FYI, I just released version 3, which removes the need to specify the individual targets. Instead, it'll scan everything in the index store, making it simpler to use for larger projects. If you're using Bazel, that's now supported out of the box. https://github.com/peripheryapp/periphery/wiki/3.0-Migration-Guide

**Comment by the301sparton** [comment]
Hey! Here at intuit we aree usie the [unused.rb](https://github.com/PaulTaykalo/swift-scripts) from time to time.
This can be used as a CLI tool or can even be integrated in xcode.

**Comment by marioelo** [comment]
I have some experience with [Periphery](https://github.com/peripheryapp/periphery) and it works great, sometimes it requires a bit of analysis but my overall experience is great

The downside is that it runs on the terminal so you should check manually the generated report inside the terminal or pipe it into a file `periphery scan > periphery-report.txt`

**Comment by jpsim** [comment]
I want to briefly chime in here to say that Periphery has gotten way better since the last time I looked into it. Notably, it's improved its performance and added first-class bazel support, and I can wholeheartedly recommend it. We've been using it at Ramp for about two months now, running all our PRs against a baseline generated from our main branch as a non-blocking informational only GH comment on PRs.

For a while I built and maintained custom tooling at Lyft to detect and delete dead code, and found it to have better performance and accuracy than Periphery, and I had started going through the motions to open source that and integrate it into SwiftLint, but I don't see a compelling reason to continue on that path now that Periphery has addressed the concerns I had with it in the past.

Great job @ileitch !

_PS: We do see Periphery crash fairly often (exit codes 133, 134, 139), and suspect that there's some data race somewhere, but running with ASan + TSan haven't pinpointed the root cause yet, so we just retry those exit codes on CI._

**Reply by ileitch** [reply]
This is great to hear, thanks @jpsim! I'd love to fix those crashes though, are you seeing crash reports being generated in Library/Logs/DiagnosticReports?
